File: source/ch/cyberduck/core/serializer/impl/PlistWriter.cs
// 
// Copyright (c) 2010-2013 Yves Langisch. All rights reserved.
// http://cyberduck.ch/
// 
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// Bug fixes, suggestions and comments should be sent to:
// yves@cyberduck.ch
// 

using System.Xml;
using ch.cyberduck.core;
using ch.cyberduck.core.serializer;
using java.util;
using Collection = java.util.Collection;

namespace Ch.Cyberduck.Core.Serializer.Impl
{
    public class PlistWriter : Writer
    {
        public void write(Collection collection, ch.cyberduck.core.Local file)
        {
            XmlDocument plistDocument = GetPlistDocument();
            XmlNode plistElement = plistDocument.LastChild;
            XmlNode arrayElement = plistDocument.CreateElement("array");
            plistElement.AppendChild(arrayElement);

            for (Iterator i = collection.iterator(); i.hasNext();)
            {
                Serializable s = (Serializable) i.next();
                XmlNode impNode = plistDocument.ImportNode((XmlNode) s.serialize(SerializerFactory.get()), true);
                arrayElement.AppendChild(impNode);
            }
            plistDocument.Save(file.getAbsolute());
        }

        public void write(Serializable bookmark, ch.cyberduck.core.Local file)
        {
            XmlDocument plistDocument = GetPlistDocument();
            XmlNode plistElement = plistDocument.LastChild;

            XmlNode impNode = plistDocument.ImportNode((XmlNode) bookmark.serialize(SerializerFactory.get()), true);
            plistElement.AppendChild(impNode);

            plistDocument.Save(file.getAbsolute());
        }

        private static XmlDocument GetPlistDocument()
        {
            XmlDocument xmlDoc = new XmlDocument();
            XmlDeclaration declaration = xmlDoc.CreateXmlDeclaration("1.0", "UTF-8", string.Empty);
            XmlNode plist = xmlDoc.CreateElement("plist");
            xmlDoc.AppendChild(declaration);
            //xmlDoc.AppendChild(type);
            xmlDoc.AppendChild(plist);

            return xmlDoc;
        }
    }
}

File: importer/src/main/java/ch/cyberduck/core/serializer/impl/jna/PlistWriter.java
package ch.cyberduck.core.serializer.impl.jna;

/*
 * Copyright (c) 2009 David Kocher. All rights reserved.
 * http://cyberduck.ch/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * Bug fixes, suggestions and comments should be sent to:
 * dkocher@cyberduck.ch
 */

import ch.cyberduck.binding.foundation.NSDictionary;
import ch.cyberduck.binding.foundation.NSMutableArray;
import ch.cyberduck.core.Local;
import ch.cyberduck.core.Serializable;
import ch.cyberduck.core.SerializerFactory;
import ch.cyberduck.core.exception.AccessDeniedException;
import ch.cyberduck.core.exception.LocalAccessDeniedException;
import ch.cyberduck.core.serializer.Writer;

import java.util.Collection;

public class PlistWriter<S extends Serializable> implements Writer<S> {

    @Override
    public void write(final Collection<S> collection, final Local file) throws AccessDeniedException {
        NSMutableArray list = NSMutableArray.array();
        for(S bookmark : collection) {
            list.addObject(bookmark.<NSDictionary>serialize(SerializerFactory.get()));
        }
        if(!list.writeToFile(file.getAbsolute())) {
            throw new LocalAccessDeniedException(String.format("Cannot create file %s", file.getAbsolute()));
        }
    }

    @Override
    public void write(final S item, final Local file) throws AccessDeniedException {
        if(!item.<NSDictionary>serialize(SerializerFactory.get()).writeToFile(file.getAbsolute())) {
            throw new LocalAccessDeniedException(String.format("Cannot create file %s", file.getAbsolute()));
        }
    }
}